### PR TITLE
basic pipeline implementation

### PR DESCRIPTION
**Pull request by jonsequitur**
The idea here is to form the foundation for a number of composable operations, such as:

* Publishing events to monitor catchup progress 
* Catching exceptions
* Logging, timing, etc.
* Managing persistence and transactions as aspects rather than as functionality of the projector itself

### CODE REVIEW

File: Alluvial.Tests/PipelineTests.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+using FluentAssertions;
+using NUnit.Framework;
+
+namespace Alluvial.Tests
+{
+    [TestFixture]
+    public class PipelineTests
+    {
+        [Test]
+        public async Task An_aggregator_can_be_short_circuited_using_Before_and_not_calling_next()
```
Comment (nathanboktae) [inline Alluvial.Tests/PipelineTests.cs]: Before() isn't actually called here

File: Alluvial.Tests/PipelineTests.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            var wasCalled = false;
+
+            var aggregator = Aggregator.Create<BalanceProjection, IDomainEvent>((projection, events) => { wasCalled = true; })
+                                       .Pipeline((projection, events, next) => projection);
+
+            var balanceProjection = new BalanceProjection();
+            var returnedProjection = aggregator.Aggregate(balanceProjection, null);
+
+            wasCalled.Should().BeFalse();
+            balanceProjection.Should().BeSameAs(returnedProjection);
```
Comment (nathanboktae) [inline Alluvial.Tests/PipelineTests.cs]: No .net equivalent for Sinon.js?

File: Alluvial.Tests/PipelineTests.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                    }
+                });
+
+            var balanceProjection = new BalanceProjection();
+            var returnedProjection = aggregator.Aggregate(balanceProjection, null);
+
+            balanceProjection.Should().BeSameAs(returnedProjection);
+        }
+    }
+}
```
Comment (nathanboktae) [inline Alluvial.Tests/PipelineTests.cs]: A test for chainability? If I read that correctly it chains processing of each event.

Comment (jonsequitur) [inline Alluvial.Tests/PipelineTests.cs]: I'll add tests for longer compositions.

Also, it's not processing per event, it's processing per batch of events.

File: Alluvial.Tests/PipelineTests.cs (hunk 2 above)
Comment (jonsequitur) [inline Alluvial.Tests/PipelineTests.cs]: This seems simple enough.

File: Alluvial.Tests/PipelineTests.cs (hunk 1 above)
Comment (jonsequitur) [inline Alluvial.Tests/PipelineTests.cs]: These started as an overload of Before and I forgot to rename the tests.

**Pull request merged**